### PR TITLE
feat: rewrite inference action in python

### LINKED ISSUE #22: Refactor inference to use a python script instead of bash
Currently, `predict.yml` uses several shell commands to:
- split the input CSV into a slice for the worker to look at
- call the ersilia model hub
- fetch and run the specified model
- upload predictions to S3

All of this can be done in python efficiently, in a manner which will allow us to better maintain and test the code. We want to write this python script and replace the bash with a call to this script..

### PR DESCRIPTION
**Pull request by kartikey-vyas**
This PR replaces the bash scripting in the `predict.yml` github workflow with a python script. This allows better maintainability and more complex pre/postprocessing on the predictions that come out of ersilia models.

### CODE REVIEW

File: .github/workflows/predict.yml [hunk 1]
```diff
@@ -18,109 +18,62 @@ on:
...
-          for (( i=1; i<$denominator; i++ ))
-          do
-            partition=$(printf "partition_%04d.csv" $i)
-            sed -i '1i smiles' "$partition"
-          done
-
-          # List files in dir
-          ls
+          shopt -s extglob
+          rm -rf !(scripts)
```
Comment (kartikey-vyas) [inline .github/workflows/predict.yml]: shouldn't need to do this, plus we want to be able to use the module we're writing

File: .github/workflows/predict.yml [hunk 2]
```diff
@@ -18,109 +18,62 @@ on:
...
-          numerator=${{ inputs.numerator }}
-          index=$((numerator - 1))
-          source activate
-          echo "Sample model id selected: ${{ inputs.model-id }}"
-          ersilia fetch ${{ inputs.model-id }}
-          ersilia serve ${{ inputs.model-id }}
-          echo "${{ inputs.model-id }} succesfully fetched and served"
-          ersilia api -i $(printf "partition_%04d.csv" $index) -o ../${{ inputs.SHA }}$(printf "_%04d.csv" $numerator)
+          python -m pip install --upgrade pip
+          pip install boto3 pandas ersilia
```
Comment (kartikey-vyas) [inline .github/workflows/predict.yml]: use `make install` instead

File: scripts/generate_predictions_with_aws_and_ersilia.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    
+    # Determine input filename based on sample-only flag
+    input_filename = f"reference_library_{sample_only}.csv" if sample_only else "reference_library.csv"
+    
+    # Fetch input data from S3
+    read_input_from_s3(bucket_name, input_filename, input_filename)
+    
+    # Generate predictions and save locally
+    output_path_template = f"../{sha}_{{numerator:04d}}.csv"
+    predictions = generate_predictions(input_filename, output_path_template, model_id, sha, numerator)
```
Comment (kartikey-vyas) [inline scripts/generate_predictions_with_aws_and_ersilia.py]: Doesn't look like you partition the input here? The objective of the original bash was to divide the input CSV into equal batches according to the number of workers, then each worker only processes a small batch of the reference library. This code just uses the numerator to name the output file, but it doesn't segment the input, so every worker would compute the full 2M rows.

File: .github/workflows/predict.yml (hunk 2 above)
Comment (kabilank6) [reply .github/workflows/predict.yml]: https://github.com/ersilia-os/model-inference-pipeline/pull/29/commits/e54b85b0574a66e77df0a6b13874e3828c763243 let me know if any concerns with this

File: scripts/generate_predictions_with_aws_and_ersilia.py (hunk 3 above)
Comment (kabilank6) [reply scripts/generate_predictions_with_aws_and_ersilia.py]: Let me have a look

File: .github/workflows/predict.yml (hunk 1 above)
Comment (kabilank6) [reply .github/workflows/predict.yml]: https://github.com/ersilia-os/model-inference-pipeline/pull/29/commits/577990f483ef2d8c1d4cd17637514ef170fc31f0

File: .github/workflows/predict.yml (hunk 2 above)
Comment (kabilank6) [reply .github/workflows/predict.yml:75]: Made a commit for this, but perhaps let's resolve the pandas/numpy issues here (https://github.com/ersilia-os/model-inference-pipeline/actions/runs/10233796760) first?

File: scripts/generate_predictions_with_aws_and_ersilia.py (hunk 3 above)
Comment (kabilank6) [reply scripts/generate_predictions_with_aws_and_ersilia.py]: Made a commit for this, but perhaps let's resolve the pandas/numpy issues here (https://github.com/ersilia-os/model-inference-pipeline/actions/runs/10233796760) first?

**Pull request merged**